spring-integration-core/src/main/java/org/springframework/integration/gateway/GatewayProxyFactoryBean.java

```diff
@@ -484,13 +486,17 @@ protected Object doInvoke(MethodInvocation invocation, boolean runningOnCallerTh
 	@Nullable
 	private Object invokeGatewayMethod(MethodInvocation invocation, boolean runningOnCallerThread) {
 		if (!this.initialized) {
-			this.afterPropertiesSet();
+			afterPropertiesSet();
 		}
 		Method method = invocation.getMethod();
 		MethodInvocationGateway gateway = this.gatewayMap.get(method);
 		Class<?> returnType = method.getReturnType();
-		boolean shouldReturnMessage = Message.class.isAssignableFrom(returnType)
-				|| hasReturnParameterizedWithMessage(method, runningOnCallerThread);
+		if (gateway.isReturnTypeMessage == null) {
```

Comment:
Package protected to avoid synthetic access method; but why is this a field there anyway - it's only used here.

Reply:
Correct. I just don't see reason to have an overhead parsing generics every single method call.
So, this is lazy loaded, but cached static state for the method. It is going to be definitely the same for each method invocation.

Do you mean that `private` field access is always wrapped to some generated setter call?

Reply:
Ah - makes sense, yes.

Yes, with a private field, the compiler has to generate synthetic static getters and setters; with package protected it uses simple GETFIELD and PUTFIELD operators.